### PR TITLE
nixos/security/ca: add support for hashed certificates

### PR DESCRIPTION
**Pull request by qubitnano**
Fixes https://github.com/NixOS/nixpkgs/issues/230575

Some applications, such as steam games, look for individual hashed certificates in /etc/ssl/certs

Tested with Drawful 2

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
Whats left to do here?

**Review by lukegb — APPROVED** [review]

File: nixos/modules/security/ca.nix [hunk 1]
```diff
@@ -35,6 +35,16 @@ in
       certificates themselves. This can have security consequences depending on your usecases
     '';
 
+    security.pki.useHashedCertificates = mkOption {
+      type = types.bool;
+      default = false;
+      example = true;
+      description = ''
+        Some applications, such as steam games, look for individual hashed certificates
```
Comment (SuperSandro2000) [inline nixos/modules/security/ca.nix:43]:
```suggestion
        Some applications, such as some Steam games, look for individual hashed certificates
```

File: nixos/modules/security/ca.nix [hunk 2]
```diff
@@ -86,7 +96,9 @@ in
 
   };
 
-  config = mkIf cfg.installCACerts {
+  config = mkMerge [
+
+  (mkIf cfg.installCACerts {
```
Comment (SuperSandro2000) [inline nixos/modules/security/ca.nix:101]:
```suggestion
  config = mkMerge [
    (mkIf cfg.installCACerts {
```
and so forth

File: nixos/modules/security/ca.nix [hunk 3]
```diff
@@ -100,6 +112,18 @@ in
...
+  })
+
+  (mkIf cfg.useHashedCertificates {
+    # Link all certificates in the hashed output to /etc/ssl/certs
+    environment.etc = let
+      hashed = "${pkgs.cacert.hashed}/etc/ssl/certs";
+    in lib.mapAttrs' (name: _: {
+      name = "ssl/certs/${name}";
+      value = { source = "${hashed}/${name}"; };
+    }) (builtins.readDir hashed);
```
Comment (SuperSandro2000) [inline nixos/modules/security/ca.nix:124]: Isn't this causing IFD? Could we just symlink the directory instead?

File: nixos/modules/security/ca.nix [hunk 4]
```diff
@@ -100,6 +112,18 @@ in
     # P11-Kit trust source.
     environment.etc."ssl/trust-source".source = "${cacertPackage.p11kit}/etc/ssl/trust-source";
 
-  };
+  })
+
+  (mkIf cfg.useHashedCertificates {
+    # Link all certificates in the hashed output to /etc/ssl/certs
+    environment.etc = let
+      hashed = "${pkgs.cacert.hashed}/etc/ssl/certs";
```
Comment (SuperSandro2000) [inline nixos/modules/security/ca.nix:120]: I wonder if we should create an option for the package

File: pkgs/data/misc/cacert/default.nix [hunk 5]
```diff
@@ -60,19 +60,21 @@ stdenv.mkDerivation rec {
...
   src = certdata;
 
-  outputs = [ "out" "unbundled" "p11kit" ];
+  outputs = [ "out" "unbundled" "p11kit" "hashed" ];
 
   nativeBuildInputs = [ buildcatrust ];
 
   buildPhase = ''
     mkdir unbundled
+    mkdir hashed
```
Comment (SuperSandro2000) [inline pkgs/data/misc/cacert/default.nix:69]:
```suggestion
    mkdir hashed unbundled
```

File: nixos/modules/programs/steam.nix [hunk 6]
```diff
@@ -201,6 +201,9 @@ in {
 
     hardware.steam-hardware.enable = true;
 
+    # Some games require hashed certificates
```
Comment (SuperSandro2000) [inline nixos/modules/programs/steam.nix:204]:
```suggestion
    # Some games (eg. Drawful 2, Jackbox Party Pack 2, Crusader Kings 3) require hashed certificates
```

**Comment by seirl** [comment]
@qubitnano hello, I originally filed the bug that your PR was attempting to fix. Can you explain why you decided to close the PR?

**Comment by qubitnano** [comment]
I couldn't find a solution that didn't use IFD. You could symlink the new hashed folder to /etc/ssl/certs, but that removes ca-certificates.crt and ca-bundle.crt, and I couldn't find an alternative to builtins.readDir (maybe symlinkJoin?) that symlinked each cert from that folder individually while keeping /etc/ssl/certs writable and the default bundles. But if anyone else wants to take a look feel free.

**Comment by seirl** [comment]
@qubitnano would you like to at least get your work on the `cacert` package merged in, since that is independent to the problem you mention?
Once the new "hashed" output is available in the package itself, we can try to come up with something that works for the NixOS module.

**Comment by qubitnano** [comment]
Sure
https://github.com/NixOS/nixpkgs/pull/370023

**Pull request closed without merging**